### PR TITLE
perf(clustering): use privileged worker for control plane connection

### PR DESCRIPTION
**Pull request by bungle**
### Summary

Data plane's connection to control plane is moved to a privileged worker process, including:
- maintaining websocket (wrpc) connection and data transfer
- decompression of received data
- json decoding of the received data
- validation and flattening of received data
- inserting data to lmdb

(so that these won't affect latencies / rps on proxy workers)

### CODE REVIEW

File: kong/clustering/data_plane.lua [hunk 1]
```diff
@@ -60,7 +61,7 @@ function _M:init_worker(plugins_list)
 
   self.plugins_list = plugins_list
 
-  if ngx.worker.id() == 0 then
+  if process.type() == "privileged agent" then
```
Comment (chronolaw) [inline kong/clustering/data_plane.lua:64]: Could we define a help function like `is_dp_worker_process` in `clusteringutils.lua`,?
then we will remove this duplicated code `process.type() == "privileged agent"`.

File: kong/init.lua [hunk 2]
```diff
@@ -667,78 +676,80 @@ function Kong.init_worker()
...
+        declarative.get_current_hash() ~= nil then
+        -- if there is no declarative config set and a config is present in LMDB,
+        -- just build the router and plugins iterator
+        ngx_log(ngx_INFO, "found persisted lmdb config, loading...")
+        local ok, err = declarative_init_build()
+        if not ok then
+          stash_init_worker_error("failed to initialize declarative config: " .. err)
+          return
+        end
+      elseif declarative_entities then
```
Comment (chronolaw) [inline kong/init.lua:702]: Add a blank line for style reason.

File: kong/init.lua [hunk 3]
```diff
@@ -667,78 +676,80 @@ function Kong.init_worker()
...
-      return
-    end
-
-    if not has_declarative_config(kong.configuration) and
-      declarative.get_current_hash() ~= nil then
-      -- if there is no declarative config set and a config is present in LMDB,
-      -- just build the router and plugins iterator
-      ngx_log(ngx_INFO, "found persisted lmdb config, loading...")
-      local ok, err = declarative_init_build()
+  if process.type ~= "privileged agent" then
```
Comment (chronolaw) [inline kong/init.lua:679]: It is a huge `if-then-end` block, could we use a simple `goto` to keep the code clean?

**Comment by dndx** [comment]
Can we measure the performance gain of this change @ADD-SP ?

File: kong/init.lua [hunk 4]
```diff
@@ -755,7 +772,7 @@ end
 
 
 function Kong.exit_worker()
-  if kong.configuration.role ~= "control_plane" and ngx.worker.id() == 0 then
+  if process.type ~= "privileged agent" and kong.configuration.role ~= "control_plane" and worker_id() == 0 then
```
Comment (chronolaw) [inline kong/init.lua:775]: should use `not is_dp_worker_process()`

Comment (bungle) [reply kong/init.lua:775]: @chronolaw I tried it, requiring "kong.clustering.utils"  causes errors on `init.lua`.

So either we need to move it else where or just keep it like it is.

File: kong/init.lua [hunk 5]
```diff
@@ -674,6 +683,13 @@ function Kong.init_worker()
 
   kong.db:set_events_handler(worker_events)
 
+  if process.type == "privileged agent" then
```
[the remark this answers is not in the document]
Comment (bungle) [reply kong/init.lua:686]: @chronolaw I tried it, requiring "kong.clustering.utils"  causes errors on `init.lua`.

So either we need to move it else where or just keep it like it is.

[the remark this answers is not in the document]
Comment (chronolaw) [reply kong/init.lua:686]: OK, thanks.

File: kong/init.lua (hunk 4 above)
Comment (chronolaw) [reply kong/init.lua:775]: OK, thanks.

File: kong/init.lua (hunk 5 above)
Comment (chronolaw) [inline kong/init.lua:686]: should be `process.type()`

File: kong/init.lua (hunk 4 above)
Comment (chronolaw) [inline kong/init.lua:775]: should be `process.type()`

File: kong/init.lua [hunk 6]
```diff
@@ -795,7 +812,7 @@ end
 
 
 function Kong.exit_worker()
-  if kong.configuration.role ~= "control_plane" and ngx.worker.id() == 0 then
+  if process.type() ~= "privileged agent" and kong.configuration.role ~= "control_plane" and worker_id() == 0 then
```
Comment (StarlightIbuki) [inline kong/init.lua:815]: Is there any reason we don't use privileged workers to manage plugin servers?

Comment (bungle) [reply kong/init.lua:815]: @StarlightIbuki I wanted to skip it for now as I don't understand all the consequences of it. E.g. is it ok to have plugin servers managed with master process rights. I suggest we do this in steps. We haven't used privileged workers before, so making too many changes at once feels too risky. We can come back to that later.

**Review by chronolaw — APPROVED** [review]

**Review by flrgh — APPROVED** [review]
Great work! I think the privileged worker process is something we should utilize more and more in the future for these types of tasks.

**Review by StarlightIbuki — APPROVED** [review]

**Comment by bungle** [comment]
> Great work! I think the privileged worker process is something we should utilize more and more in the future for these types of tasks.

Yes, next obvious target is Kong EE Vitals.

**Pull request merged**